src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java

```diff
@@ -150,20 +142,28 @@ public static float cosineSimilarity(List<Number> queryVector, KNNVectorScriptDo
      */
     public static float cosinesimil(float[] queryVector, float[] inputVector) {
         requireEqualDimension(queryVector, inputVector);
-        float dotProduct = 0.0f;
-        float normQueryVector = 0.0f;
-        float normInputVector = 0.0f;
-        for (int i = 0; i < queryVector.length; i++) {
-            dotProduct += queryVector[i] * inputVector[i];
-            normQueryVector += queryVector[i] * queryVector[i];
-            normInputVector += inputVector[i] * inputVector[i];
+        int numZeroInInput = 0;
+        int numZeroInQuery = 0;
+        float cosine = 0.0f;
+        for (int i = 0; i < inputVector.length; i++) {
+            if (inputVector[i] == 0) {
+                numZeroInInput++;
+            }
+
+            if (queryVector[i] == 0) {
+                numZeroInQuery++;
+            }
         }
-        float normalizedProduct = normQueryVector * normInputVector;
-        if (normalizedProduct == 0) {
+        if (numZeroInInput == inputVector.length || numZeroInQuery == queryVector.length) {
+            return cosine;
+        }
+        try {
+            cosine = VectorUtil.cosine(queryVector, inputVector);
+        } catch (Exception e) {
```

Comment:
Can we avoid catching all exceptions here or catch a more specific exception?

Reply:
@jmazanec15 Sure I will change it